crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+    };
+    use reth_interfaces::test_utils::TestConsensus;
+    use reth_primitives::{hex_literal::hex, proofs::EMPTY_ROOT, ChainSpecBuilder, H256, MAINNET};
+    use reth_provider::{
+        execution_result::ExecutionResult, insert_block, test_utils::blocks, BlockExecutor,
+    };
+
+    struct TestFactory {
+        exec_result: Arc<Mutex<Vec<ExecutionResult>>>,
+        chain_spec: Arc<ChainSpec>,
+    }
+
+    impl TestFactory {
+        fn new(chain_spec: Arc<ChainSpec>) -> Self {
+            Self { exec_result: Arc::new(Mutex::new(Vec::new())), chain_spec }
+        }
+
+        fn extend(&self, exec_res: Vec<ExecutionResult>) {
+            self.exec_result.lock().extend(exec_res.into_iter());
+        }
+    }
+
+    struct TestExecutor(Option<ExecutionResult>);
+
+    impl<SP: StateProvider> BlockExecutor<SP> for TestExecutor {
+        fn execute(
+            &mut self,
+            _block: &reth_primitives::Block,
+            _total_difficulty: reth_primitives::U256,
+            _senders: Option<Vec<reth_primitives::Address>>,
+        ) -> Result<ExecutionResult, ExecError> {
+            self.0.clone().ok_or(ExecError::VerificationFailed)
+        }
+
+        fn execute_and_verify_receipt(
+            &mut self,
+            _block: &reth_primitives::Block,
+            _total_difficulty: reth_primitives::U256,
+            _senders: Option<Vec<reth_primitives::Address>>,
+        ) -> Result<ExecutionResult, ExecError> {
+            self.0.clone().ok_or(ExecError::VerificationFailed)
+        }
+    }
+
+    impl ExecutorFactory for TestFactory {
+        type Executor<T: StateProvider> = TestExecutor;
+
+        fn with_sp<SP: StateProvider>(&self, _sp: SP) -> Self::Executor<SP> {
+            let exec_res = self.exec_result.lock().pop();
+            TestExecutor(exec_res)
+        }
+
+        fn chain_spec(&self) -> &ChainSpec {
+            self.chain_spec.as_ref()
+        }
+    }
+
+    type TestExternals = Externals<Arc<Env<WriteMap>>, TestConsensus, TestFactory>;
+
+    fn externals(exec_res: Vec<ExecutionResult>) -> TestExternals {
+        let db = create_test_rw_db();
+        let consensus = TestConsensus::default();
+        let chain_spec = Arc::new(
+            ChainSpecBuilder::default()
+                .chain(MAINNET.chain)
+                .genesis(MAINNET.genesis.clone())
+                .shanghai_activated()
+                .build(),
+        );
+        let executor_factory = TestFactory::new(chain_spec.clone());
+        executor_factory.extend(exec_res);
+
+        Externals { db, consensus, executor_factory, chain_spec }
+    }
+
+    fn setup(externals: &TestExternals) {
+        // insert genesis to db.
+
+        let mut genesis = blocks::genesis();
+        genesis.header.header.number = 10;
+        genesis.header.header.state_root = EMPTY_ROOT;
+        let tx_mut = externals.db.tx_mut().unwrap();
+
+        tx_mut.put::<tables::AccountsTrie>(EMPTY_ROOT, vec![0x80]).unwrap();
```

Comment:
nit is this needed

Reply:
No, will remove it, it will be automatically added.